File: backend(Spring Boot)/src/main/java/com/peergrid/backend/controller/WebinarController.java
package com.peergrid.backend.controller;

import com.peergrid.backend.entity.Webinar;
import com.peergrid.backend.entity.User;
import com.peergrid.backend.repository.WebinarRepository;
import com.peergrid.backend.repository.UserRepository;
import com.peergrid.backend.service.EmailService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.security.Principal;
import java.time.LocalDateTime;
import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/webinars")
@CrossOrigin(origins = "http://localhost:5173", allowedHeaders = "*", allowCredentials = "true")
public class WebinarController {

    @Autowired
    private WebinarRepository webinarRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private EmailService emailService;

    // Get all upcoming webinars
    @GetMapping
    public List<Webinar> getAllWebinars() {
        return webinarRepository.findByScheduledTimeAfterOrderByScheduledTimeAsc(LocalDateTime.now());
    }

    // Create a new webinar (Host only? Anyone can host? Let's assume anyone logged in)
    @PostMapping
    public ResponseEntity<?> createWebinar(@RequestBody Webinar webinar, Principal principal) {
        if (principal == null) {
            return ResponseEntity.status(401).body("Unauthorized");
        }

        User host = userRepository.findByEmail(principal.getName())
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        webinar.setHost(host);
        // Ensure webinar time is in future?
        if (webinar.getScheduledTime().isBefore(LocalDateTime.now())) {
            return ResponseEntity.badRequest().body("Cannot schedule webinar in the past.");
        }

        Webinar saved = webinarRepository.save(webinar);
        return ResponseEntity.ok(saved);
    }

    // Register for a webinar
    @PostMapping("/{id}/register")
    public ResponseEntity<?> registerForWebinar(@PathVariable Integer id, Principal principal) {
        if (principal == null) {
            return ResponseEntity.status(401).body("Unauthorized");
        }

        User user = userRepository.findByEmail(principal.getName())
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        Webinar webinar = webinarRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Webinar not found"));
        
        if (webinar.getRegisteredUserIds().contains(user.getId())) {
             return ResponseEntity.badRequest().body("Already registered.");
        }

        BigDecimal cost = webinar.getCost();
        if (user.getGridPoints().compareTo(cost) < 0) {
            return ResponseEntity.badRequest().body("Insufficient Grid Points.");
        }

        // Deduct points
        user.setGridPoints(user.getGridPoints().subtract(cost));
        userRepository.save(user);

        // Add user to webinar
        webinar.getRegisteredUserIds().add(user.getId());
        webinarRepository.save(webinar);

        // Send Email
        emailService.sendWebinarRegistrationEmail(
            user.getEmail(), 
            user.getName(), 
            webinar.getTitle(), 
            webinar.getScheduledTime().toString(), 
            webinar.getMeetingLink() != null ? webinar.getMeetingLink() : "Link will be shared soon"
        );

        return ResponseEntity.ok(Map.of("message", "Registered successfully!", "newBalance", user.getGridPoints()));
    }
}

File: backend(.NET)/Controllers/WebinarController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using PeerGrid.Backend.Data;
using PeerGrid.Backend.Models;
using PeerGrid.Backend.Services;
using System.Security.Claims;

namespace PeerGrid.Backend.Controllers
{
    [Route("api/webinars")]
    [ApiController]
    public class WebinarController : ControllerBase
    {
        private readonly ApplicationDbContext _context;
        private readonly IEmailService _emailService;

        public WebinarController(ApplicationDbContext context, IEmailService emailService)
        {
            _context = context;
            _emailService = emailService;
        }

        // GET: api/webinars
        [HttpGet]
        // GET: api/webinars
        [HttpGet]
        public async Task<ActionResult<IEnumerable<object>>> GetWebinars()
        {
            var webinars = await _context.Webinars
                .Include(w => w.Host)
                .Include(w => w.RegisteredUsers) 
                .Where(w => w.ScheduledTime > DateTime.Now)
                .OrderBy(w => w.ScheduledTime)
                .ToListAsync();

            var result = webinars.Select(w => new {
                w.Id,
                w.Title,
                w.Description,
                w.HostId,
                w.Host, // Includes Host details
                w.ScheduledTime,
                w.DurationMinutes,
                w.Cost,
                w.MeetingLink,
                RegisteredUserIds = w.RegisteredUsers.Select(u => u.Id).ToList()
            });

            return Ok(result);
        }

        // POST: api/webinars
        [HttpPost]
        [Authorize]
        public async Task<ActionResult<Webinar>> CreateWebinar(Webinar webinar)
        {
            var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)?.Value);
            var user = await _context.Users.FindAsync(userId);

            if (user == null) return NotFound("User not found");

            webinar.HostId = userId;
            // Ensure time is future
            if (webinar.ScheduledTime < DateTime.Now)
            {
                return BadRequest("Cannot schedule webinar in the past.");
            }

            _context.Webinars.Add(webinar);
            await _context.SaveChangesAsync();

            return CreatedAtAction("GetWebinars", new { id = webinar.Id }, webinar);
        }

        // POST: api/webinars/5/register
        [HttpPost("{id}/register")]
        [Authorize]
        public async Task<IActionResult> RegisterForWebinar(int id)
        {
            var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)?.Value);
            var user = await _context.Users.FindAsync(userId);
            if (user == null) return NotFound("User not found");

            var webinar = await _context.Webinars.Include(w => w.RegisteredUsers).FirstOrDefaultAsync(w => w.Id == id);
            if (webinar == null) return NotFound("Webinar not found");

            if (webinar.RegisteredUsers.Any(u => u.Id == userId))
            {
                return BadRequest("Already registered.");
            }

            if (user.GridPoints < webinar.Cost)
            {
                return BadRequest("Insufficient Grid Points.");
            }

            // Deduct Points
            user.GridPoints -= webinar.Cost;
            
            // Add to webinar
            webinar.RegisteredUsers.Add(user);

            await _context.SaveChangesAsync();

            // Send Email
            _emailService.SendWebinarRegistrationEmail(
                user.Email,
                user.Name,
                webinar.Title,
                webinar.ScheduledTime.ToString(),
                webinar.MeetingLink ?? "Link will be shared soon"
            );

            return Ok(new { message = "Registered successfully!", newBalance = user.GridPoints });
        }
    }
}
